mqtt/mqtt-bridge/src/pump/builder.rs

```diff
@@ -0,0 +1,161 @@
+use std::convert::TryInto;
+
+use tokio::sync::mpsc;
+
+use crate::{
+    bridge::BridgeError,
+    client::{MqttClient, MqttClientConfig},
+    messages::{MessageHandler, TopicMapper},
+    persist::{PublicationStore, WakingMemoryStore},
+    settings::TopicRule,
+    upstream::{
+        LocalRpcHandler, LocalUpstreamHandler, LocalUpstreamPumpEventHandler, RemoteRpcHandler,
+        RemoteUpstreamHandler, RemoteUpstreamPumpEventHandler,
+    },
+};
+
+use super::{MessagesProcessor, Pump, PumpHandle};
+
+pub type PumpsResult = Result<
+    (
+        Pump<LocalUpstreamHandler<WakingMemoryStore>, LocalUpstreamPumpEventHandler>,
+        Pump<RemoteUpstreamHandler<WakingMemoryStore>, RemoteUpstreamPumpEventHandler>,
+    ),
+    BridgeError,
+>;
+
+pub struct Builder {
+    local: PumpBuilder,
+    remote: PumpBuilder,
+    store: Box<dyn Fn() -> PublicationStore<WakingMemoryStore>>,
+}
+
+impl Default for Builder {
+    fn default() -> Self {
+        Self {
+            local: PumpBuilder::default(),
+            remote: PumpBuilder::default(),
+            store: Box::new(|| PublicationStore::new_memory(0)),
+        }
+    }
+}
+
+impl Builder {
+    pub fn with_local<F>(&mut self, mut apply: F) -> &mut Self
+    where
+        F: FnMut(&mut PumpBuilder),
+    {
+        apply(&mut self.local);
+        self
+    }
+
+    pub fn with_remote<F>(&mut self, mut apply: F) -> &mut Self
+    where
+        F: FnMut(&mut PumpBuilder),
+    {
+        apply(&mut self.remote);
+        self
+    }
+
+    pub fn with_store<F>(&mut self, store: F) -> &mut Self
+    where
+        F: Fn() -> PublicationStore<WakingMemoryStore> + 'static,
+    {
+        self.store = Box::new(store);
+        self
+    }
+
+    pub fn build(&mut self) -> PumpsResult {
+        let ingress_store = (self.store)();
+        let ingress_loader = ingress_store.loader();
+
+        let egress_store = (self.store)();
+        let egress_loader = egress_store.loader();
+
+        let (remote_messages_send, remote_messages_recv) = mpsc::channel(100);
+        let (local_messages_send, local_messages_recv) = mpsc::channel(100);
+
+        let (subscriptions, topic_filters) = make_topics(&self.local.rules)?;
+
+        let rpc = LocalRpcHandler::new(PumpHandle::new(remote_messages_send.clone()));
+        let messages = MessageHandler::new(ingress_store.clone(), topic_filters);
+        let handler = LocalUpstreamHandler::new(messages, rpc);
+
+        let config = self.local.client.take().expect("local client config");
+        let client = MqttClient::tls(config, handler);
```

Comment:
does local client have to be tls?

Reply:
@ancaantochi / @and-rewsmith may have more context on this

Reply:
local client should be tcp